db/kv_channel_storage.go

```diff
@@ -30,29 +30,33 @@ const (
 	kSequenceOffsetLength = 0 // disabled until we actually need it
 )
 
-// ChannelStorage implemented as an interface, to support swapping to different underlying storage model
+// ChannelStorage implemented as two interfaces, to support swapping to different underlying storage model
 // without significant refactoring.
-type ChannelStorage interface {
+type ChannelStorageReader interface {
+	// GetAllEntries returns all entries for the channel in the specified range, for all vbuckets
+	GetChanges(fromSeq base.SequenceClock, channelClock base.SequenceClock, limit int) ([]*LogEntry, error)
+}
+
+type ChannelStorageWriter interface {
 
 	// AddEntrySet adds a set of entries to the channel index
 	AddEntrySet(entries []*LogEntry) (clockUpdates base.SequenceClock, err error)
 
-	// GetAllEntries returns all entries for the channel in the specified range, for all vbuckets
-	GetChanges(fromSeq base.SequenceClock, channelClock base.SequenceClock) ([]*LogEntry, error)
-
 	// If channel storage implementation uses separate storage for log entries and channel presence,
 	// WriteLogEntry and ReadLogEntry can be used to read/write.  Useful when changeIndex wants to
 	// manage these document outside the scope of a channel.  StoresLogEntries() allows callers to
 	// check whether this is available.
 	StoresLogEntries() bool
 	WriteLogEntry(entry *LogEntry) error
+}
 
-	// For unit testing only
-	getIndexBlockForEntry(entry *LogEntry) IndexBlock
+type ChannelStorage interface {
+	ChannelStorageReader
+	ChannelStorageWriter
 }
 
-func NewChannelStorage(bucket base.Bucket, channelName string, partitions *base.IndexPartitions) ChannelStorage {
-	return NewBitFlagStorage(bucket, channelName, partitions)
+func NewChannelStorageReader(bucket base.Bucket, channelName string, partitions *base.IndexPartitions) ChannelStorageReader {
+	return NewDenseStorageReader(bucket, channelName, partitions)
```

Comment:
I see where you're coming from for something like NewDenseStorageReader - it should be (and is) returning the impl and not the interface it's implementing. 

I'm less sure about this scenario - the original motivation was to act as a single location that would need to be changed in order to swap between storage types.  In this scenario I would have thought that returning the interface is more appropriate, as it ensures that callers aren't inadvertently depending on specifics of an impl that aren't in the interface.  (having said that, after splitting into Reader/Writer, it's not really serving that purpose any more - there's only a single caller).

If you've got more details on the rationale, I'd be interested to learn more.